File: lib/base/pairs.go
package base

import rt "github.com/arnodel/golua/runtime"

func pairs(t *rt.Thread, c *rt.GoCont) (rt.Cont, error) {
	if err := c.Check1Arg(); err != nil {
		return nil, err
	}
	coll := c.Arg(0)
	next := c.Next()
	res := rt.NewTerminationWith(c, 0, true)
	err, ok := rt.Metacall(t, coll, "__pairs", []rt.Value{coll}, res)
	if ok {
		if err != nil {
			return nil, err
		}
		t.Push(next, res.Etc()...)
		return next, nil
	}
	t.Push(next, rt.FunctionValue(nextGoFunc), coll, rt.NilValue)
	return next, nil
}

File: lib/base/lua/pairs.lua
function _pairs(t)
    return next, t, nil
end

local t = {x=1, y=2, z="hello"}
local u = {}
for k, v in _pairs(t) do
    u[k] = v
end
print(u.x, u.y, u.z)
--> =1	2	hello

u = {}
for k, v in pairs(t) do
    u[k] = v .. "!"
end
print(u.x, u.y, u.z)
--> =1!	2!	hello!

local tt = {}
setmetatable(tt, {__pairs=function(x) return "what" end})
print(pairs(tt))
--> =what

print(pcall(pairs))
--> ~false\t.*value needed

setmetatable(tt, {__pairs=function(x) error("nono") end})
print(pcall(pairs, tt))
--> ~false\t.* nono
